Create Feature Labels

wiersgallak opened the pull request:
#### Summary
Adding a new section under R&D for shared processes and adding feature labels as first docs

wiersgallak committed:
Create Feature Labels

matthewbirtch commented on operations/research-and-development/Processes/Feature Labels:
```suggestion
Solution has undergone thorough validation and testing. It is feature-complete, meets quality standards, and has successfully passed security reviews. GA features are suitable for widespread production deployment and adoption, as they offer stability and reliability, with no expected changes that could disrupt functionality or scalability.
```

Just needs a period

matthewbirtch commented on operations/research-and-development/Processes/Feature Labels:
```suggestion
Solution is being considered for removal from the product due to redundancy, low usage, or compatibility issues. It is still functional but is no longer actively developed or enhanced. Only severe bugs and security issues are addressed. Users are encouraged to transition away from using this feature towards alternative solutions. 
```

matthewbirtch reviewed (dismissed):
Looks great to me. Just a few minor suggestions

wiersgallak committed:
Update operations/research-and-development/Processes/Feature Labels

Co-authored-by: Matthew Birtch <mattbirtch@gmail.com>

wiersgallak committed:
Update operations/research-and-development/Processes/Feature Labels

Co-authored-by: Matthew Birtch <mattbirtch@gmail.com>

wiersgallak committed:
Rename Feature Labels to feature-labels

wiersgallak committed:
Update SUMMARY.md

wiersgallak committed:
Create READ.me

wiersgallak committed:
Rename READ.me to README

wiggin77 commented on operations/research-and-development/processes/README:
```diff
@@ -0,0 +1,5 @@
+# R&D Processes
+
+This section outlines key processes shared between Product, Engineering, Quality Assuarnce, Security, and Product Operaions
```
```suggestion
This section outlines key processes shared between Product, Engineering, Quality Assurance, Security, and Product Operations
```

wiggin77 commented on operations/research-and-development/processes/feature-labels:
This is called ""General Availability" above.

lindalumitchell replied:
+1 to "General Availability"

wiggin77 commented on operations/research-and-development/processes/feature-labels:
Is this really a thing at Mattermost? We have no way to get PRs approved/merged unless a MM team is assigned to monitor them. It becomes a frustrating experience for community members who submit PRs because it sounds like we have a "Community Supported" program where their contributions will be reviewed.

lindalumitchell replied:
I wonder the same thing. Do we have examples of how this has worked well in the past, or is this more aspirational? Would we want to incorporate some sort of process / triggers to get our attention when community folks want to update a community-supported feature?

wiersgallak replied:
I drafted this based on us moving some integrations from Mattermost Support to "Community support". @wiggin77 or @lindalumitchell do you have any suggestions on alternative labels? Or are you advocating for no label?

esethna replied:
I would advocate for no label in this case. Community supported = "deprecated" from an internal development standpoint and I think that's the kind of certainty and clarity that would help a customer or user understand what level of support/risk to expect.

wiersgallak replied:
I've removed this for now. Let's continue to assess if we want to be more specific on labeling as such in the future.

lindalumitchell commented on SUMMARY.md:
```diff
@@ -55,6 +55,8 @@
     * [GRC](/operations/research-and-development/organization/grc.md)
     * [Product Security](/operations/research-and-development/organization/product_security.md)
     * [Security Operations](/operations/research-and-development/organization/security_operations.md)
+  * [Processes](operations/research-and-development/processes/README.md/)
```
Given that there are additional process docs below in development and release sections for example, should these new docs integrate into those sections, or might we want to unite other process docs up in this new section? I see the value of wanting to call out the cross-org-ness of the new process docs, but also think there is an argument for many of the other docs in the other process sections being relevant across the org as well.

wiersgallak replied:
@lindalumitchell would you be willing to help determine what those are and create a doc then we can work through in our committee?

lindalumitchell commented on operations/research-and-development/processes/feature-labels:
```suggestion
Solution is in active development toward General Availability. Not fully complete, but reviewed by our security team, adoption is suitable for a small set of customers behind a feature flag. Identified bugs are addressed, but no major breaking changes are anticipated. Beta is a transitional stage, meaning the solution is maturing but requires careful consideration in a full production deployment as scale and client availability may vary.
```

esarafianou replied:
"Fixed on best effort" and addressed doesn't mean the same. I think that we want to keep to "fixed on best effort" which indicates that there's no 100% guarantee.

iyampaul replied:
"best effort" and "no major breaking changes" is also only the minimum bar for a feature to reach beta, far from reaching GA.

esethna replied:
+1 to keeping the best effort statement

```suggestion
Solution is in active development towards General Availability. Not fully complete, but reviewed by our security team, adoption is suitable for a small set of customers behind a feature flag. Identified bugs are fixed on a best effort basis, and no major breaking changes are anticipated. Beta is a transitional stage, meaning the solution is maturing but requires careful consideration in a full production deployment as scale and client availability may vary.

lindalumitchell replied:
Sounds good. I wasn't familiar with the implications of "best effort", and thought that "addressed" left even more room for deciding what to fix or not, for example. But if the implications of that phrase are generally understood I'm certainly fine with keeping it.

iyampaul commented on operations/research-and-development/processes/feature-labels:
> While the feature may still be available in the current version

If the feature still exists in a supported version, in the least we need to resolve severe bugs and security issues (similar to the "Candidate for Deprecation").

wiersgallak replied:
@iyampaul Would it help to say "If the feature still exists in a customer's deployed version"

esethna replied:
Perhaps:

```suggestion
Solution is officially marked for removal from the product. It is no longer supported or actively maintained by the development team. If the feature is still in use in your deployed version, we recommend users discontinue its use and migrate to alternative functionalities.
```

iyampaul replied:
If the feature is still in their deployed version and yet the feature has been deprecated, this implies their deployed version is out of support.  Should we then encourage them to update?

marianunez replied:
> this implies their deployed version is out of support.

They could still be in an ESR for which we wouldn't consider a feature `Deprecated` in that version yet given that we will still fix security issues as needed. The recommendation here to migrate out of a feature that has been announced as Deprecated in the next version makes sense to me.

iyampaul replied:
The idea of announcing it is being deprecated in the next version while still supporting it in ESR makes sense.  Maybe we need to use a different label for a ESR-only features?

wiersgallak replied:
In my opinion, I'd prefer to not add another label. I think we could mark it deprecated as of x.xx. My hope is that for customers still on a version (like an ESR) this sparks a conversation with their CSM for a transition plan. I also hope that we don't have too many in the future that require longer-term ESR support.

wiersgallak committed:
Update operations/research-and-development/processes/README

Co-authored-by: Doug Lauder <wiggin77@warpmail.net>

wiersgallak commented on operations/research-and-development/processes/feature-labels:
```suggestion
### General Availablity 
```

wiersgallak committed:
Update operations/research-and-development/processes/feature-labels

wiersgallak committed:
Update operations/research-and-development/processes/feature-labels

Co-authored-by: Eric Sethna <eric@mattermost.com>

wiersgallak committed:
Update operations/research-and-development/processes/feature-labels

Co-authored-by: Eric Sethna <eric@mattermost.com>

wiersgallak committed:
Update feature-labels

Removed "community supported"

wiersgallak commented:
@wiggin77 @iyampaul @lindalumitchell I've addressed feedback and submitted for you're review again if you'd like to give it another review with input addressed.

iyampaul commented on operations/research-and-development/processes/feature-labels:
> Users are encouraged to transition away from using this feature towards alternative solutions. 
Comment

Describing this stage as a "Candidate for Deprecation" (to me) implies we haven't yet decided to cut it, but then by encouraging users to stop using the feature it will contribute to the "low usage" further justifying the deprecation. So by saying a feature is a "Candidate for Deprecation" we're just warning people it _will_ happen - why not just say it's being deprecated then set a date or release right away?

wiersgallak replied:
@iyampaul We typically use the forums to detail out timelines and context. I could see this label as a pre-warning and we can link in the documentation to the details. We want to prevent surprises for customers, this could be a way to earn trust.

iyampaul replied:
@wiersgallak - I think my challenge is with the term `Candidate` as it means there's a chance it won't be deprecated, but in reality this is a sure thing.  I could see a better approach to this being "sunsetting" features or a similar concept.

wiersgallak replied:
@iyampaul Fair. I have removed it for this iteration and we can discuss the process before putting it into the handbook.

wiersgallak committed:
Update feature-labels

cwarnermm commented on operations/research-and-development/processes/README:
```diff
@@ -0,0 +1,5 @@
+# R&D Processes
+
+This section outlines key processes shared between Product, Engineering, Quality Assurance, Security, and Product Operations
```
```suggestion
This section outlines key processes shared between Product, Engineering, Quality Assurance, Security, and Product Operations.
```

cwarnermm commented on operations/research-and-development/processes/feature-labels:
```suggestion
Solution has undergone thorough validation and testing. It is feature-complete, meets quality standards, and has successfully passed security reviews. General Availability features are suitable for widespread production deployment and adoption, as they offer stability and reliability, with no expected changes that could disrupt functionality or scalability.
```

cwarnermm committed:
Update operations/research-and-development/processes/README

cwarnermm committed:
Update operations/research-and-development/processes/feature-labels
